pkg/controller/volume/expand/expand_controller.go

```diff
@@ -123,107 +113,146 @@ func NewExpandController(
 	}
+}
 
-	// Run volume sync work goroutine
-	go expc.syncResize.Run(stopCh)
-	// Start the pvc populator loop
-	go expc.pvcPopulator.Run(stopCh)
-	<-stopCh
+func (expc *expandController) processNextWorkItem() bool {
+	key, shutdown := expc.queue.Get()
+	if shutdown {
+		return false
+	}
+	defer expc.queue.Done(key)
+
+	err := expc.syncHandler(key.(string))
+	if err == nil {
+		expc.queue.Forget(key)
+		return true
+	}
+
+	runtime.HandleError(fmt.Errorf("%v failed with : %v", key, err))
+	expc.queue.AddRateLimited(key)
+
+	return true
 }
 
-func (expc *expandController) deletePVC(obj interface{}) {
-	pvc, ok := obj.(*v1.PersistentVolumeClaim)
-	if !ok {
-		tombstone, ok := obj.(kcache.DeletedFinalStateUnknown)
-		if !ok {
-			runtime.HandleError(fmt.Errorf("couldn't get object from tombstone %+v", obj))
-			return
-		}
-		pvc, ok = tombstone.Obj.(*v1.PersistentVolumeClaim)
-		if !ok {
-			runtime.HandleError(fmt.Errorf("tombstone contained object that is not a pvc %#v", obj))
-			return
+func (expc *expandController) syncHandler(key string) error {
+	namespace, name, err := kcache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+	pvc, err := expc.pvcLister.PersistentVolumeClaims(namespace).Get(name)
+	if errors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		klog.V(5).Infof("Error getting PVC %q (uid: %q) from informer : %v", util.GetPersistentVolumeClaimQualifiedName(pvc), pvc.UID, err)
+		return err
+	}
+
+	pv, err := getPersistentVolume(pvc, expc.pvLister)
+	if err != nil {
+		klog.V(5).Infof("Error getting Persistent Volume for PVC %q (uid: %q) from informer : %v", util.GetPersistentVolumeClaimQualifiedName(pvc), pvc.UID, err)
+		return err
+	}
+	if pv.Spec.ClaimRef == nil || pvc.Namespace != pv.Spec.ClaimRef.Namespace || pvc.Name != pv.Spec.ClaimRef.Name {
+		err := fmt.Errorf("Persistent Volume is not bound to PVC being updated : %s", util.ClaimToClaimKey(pvc))
+		klog.V(4).Infof("%v", err)
+		return err
+	}
+
+	volumeSpec := volume.NewSpecFromPersistentVolume(pv, false)
+	volumePlugin, err := expc.volumePluginMgr.FindExpandablePluginBySpec(volumeSpec)
+	if err != nil || volumePlugin == nil {
+		msg := fmt.Errorf("didn't find a plugin capable of expanding the volume; " +
+			"waiting for an external controller to process this PVC")
+		eventType := v1.EventTypeNormal
+		if err != nil {
+			eventType = v1.EventTypeWarning
 		}
+		expc.recorder.Event(pvc, eventType, events.ExternalExpanding, fmt.Sprintf("Ignoring the PVC: %v.", msg))
+		klog.V(3).Infof("Ignoring the PVC %q (uid: %q) : %v.", util.GetPersistentVolumeClaimQualifiedName(pvc), pvc.UID, msg)
+		return err
 	}
 
-	expc.resizeMap.DeletePVC(pvc)
+	return expc.expand(pvc, pv)
 }
 
-func (expc *expandController) pvcUpdate(oldObj, newObj interface{}) {
-	oldPVC, ok := oldObj.(*v1.PersistentVolumeClaim)
+func (expc *expandController) expand(pvc *v1.PersistentVolumeClaim, pv *v1.PersistentVolume) error {
+	pvc, err := util.MarkResizeInProgress(pvc, expc.kubeClient)
```

Comment:
Since there is no resizing failure condition, presence of a perpetual resizing condition means volume is still being considered for resizing and resizing is pending on the volume.

Reply:
ack